apps/src/applab/designMode.js

```diff
@@ -613,12 +613,13 @@ designMode.onDuplicate = function(element, event) {
 
 var batchChangeId = 1;
 
-designMode.changeThemeForCurrentScreen = function(prevThemeValue, themeValue) {
-  const currentScreen = $(
-    elementUtils.getPrefixedElementById(
-      getStore().getState().screens.currentScreenId
-    )
-  );
+designMode.changeThemeForCurrentScreen = function(screenElement, themeValue) {
+  const prevThemeValue =
+    screenElement.getAttribute('data-theme') ||
+    applabConstants.themeOptions[applabConstants.CLASSIC_THEME_INDEX];
+  screenElement.setAttribute('data-theme', themeValue);
```

Comment:
Really would like to see each property have know whether it's been modified (and show this in the UI!) rather than trying to infer it from the current property.

**Am I understanding correctly that we can never change theme values**, or elements created in that theme in previous student projects won't be identified as "the same", and therefore won't update when the student changes the theme?

We've used `data-*` attributes for this in the past — could be a good option here!

Reply:
Yes, as discussed in previous reviews, we decided to have limited understanding of what has in fact been modified. Even if we did track it separately, we have to deal with the question of did the user change it back to "default", at which point we'd probably "unset" the `data-modified` attribute, and the result would be similar. It's really a tricky balancing act.

As I mentioned verbally, we are looking at a model where at the time of a theme change, the student is given the option to choose between honoring customizations they've made (which makes switching between themes and changing back non-destructive) and overwriting all themeable properties with the new values (making it destructive)

Reply:
There's lots of good prior work on UX for inherited vs custom settings ([examples](https://ux.stackexchange.com/questions/75236/displaying-custom-vs-default-settings)). I've got opinions here, but in short it doesn't feel like we're implementing a known best practice for this sort of thing.  Since this decision affects user data, if we're going to change our approach we should either do it now or figure out our approach to a future upgrade.

If this whole conversation has already happened and we're definitely going with this solution, can you point us to spec or other documentation of that discussion so we can all start on the same page?

Reply:
yes, this was already discussed. this is an imperfect solution, but we can certainly evolve it as needed. Many of the "reset to default" scenarios have been discussed as potential follow-on items (either on a per-property basis or in bulk for an element or an entire screen)

@ryansloan can you share what was already agreed upon in the spec and review discussions?

Reply:
A bit more discussion in Slack here: https://codedotorg.slack.com/archives/C0SUN2SSF/p1557362993028800

Reply:
Hey, sorry missed this thread the first time. 

[This](https://codedotorg.slack.com/archives/C0T0UQPLZ/p1551486694002300) is the summary plan we were working from:

- *Changing the theme will cause all elements on the _current screen_ to update.*
- We’re going to try to build it such that any style value on any element that differs from the default will be IGNORED during the theme change, but if the element still has the default property (for the current theme), then that property will be updated according to the new theme
- Longer term: Potentially change the appearance of the icons in the design mode toolbox to match the current theme
- We’re not supporting changing themes at runtime (we’re approaching this as a design mode feature)
- Any elements created at runtime (e.g. `button('id', 'text')`) respect the theme for the current screen

[This is Poorva's spec](https://docs.google.com/document/d/1WYVcn6FB5hOufvHkz8epK4QHeB7RkyvWu4tdRDJ1iPM/edit#) (pages 1-2)

I talked with Chris about exploring options for applying to all (to override custom styles). If we're able to, I'd like to prompt the user for a decision whether or not to override customizations (if they have them) when they choose a new theme. But based on some of the wonky ways that inline styles work, I'm kind of trying to start from what our feasible options are.

Reply:
The spec and Slack discussion are thin on the edge cases we're discussing, and haven't yet convinced me that we've thought this through.

The spirit of the thing seems to be "My styles change with the stylesheet, until I customize them, and then they stay custom forever." So far we've identified two scenarios that seem to violate this idea.

1. [The test case I requested](https://github.com/code-dot-org/code-dot-org/pull/28402#discussion_r282350345): A customized style may become uncustomized just by trying out different themes on one's project.

2. [Josh's scenario at the top of this thread](https://github.com/code-dot-org/code-dot-org/pull/28402#discussion_r282299066): If we ever change a theme definition in the future, existing projects will see uncustomized styles become custom and may see custom styles become uncustomized.

I want confirmation that we're genuinely okay with these product scenarios (I'm not a fan of them).  If we are _temporarily_ okay with the product scenarios, I want to totally think through what fixing them later looks like, because I think it's going to be tough to retroactively know which styles students have customized as currently built.

The second scenario in particular seems like a footgun to me. We might be able to safely change a theme with an appropriate migration of existing projects, but what prevents us from forgetting to do such a migration?

Reply:
For #2 above, the onus will definitely be on us to add the versioning/migration if we ever decide to change a theme value from this point forward.

Reply:
Hi everyone, I'm taking this out of the PR to summarize a proposed plan. You can see it [here](https://docs.google.com/document/d/1498j0bn56eNvSsh2yhwaeH1RPrUYM2lYzHAjlXylYmU/edit#) (also sent via email)